src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunTableManager.cs

```diff
@@ -44,6 +59,24 @@ public IEnumerable<TypeDesc> GetTypesWithAvailableTypes()
             return _typesWithAvailableTypesGenerated;
```

Comment:
Should this enumerator be based on direct Metadata too, just like `GetExportedTypes`?

Reply:
Well, I think the difference is that "available types" currently come from the dependency analyzer so they're guaranteed to have correct TypeDesc's; while "exported types" are arbitrary forwards and are susceptible to various forms of mismatches and resolution issues as Michal often points out.

According to my understanding there's no real gain in switching over the internal representation of available types from TypeDesc's to tokens; if you're aware of any, please let me know, otherwise it seems to me that, for available types, the code is generally simpler when working with TypeDesc's, If you know a wicked counterexample that invalidates my assessment, I'd certainly love to hear it sooner rather than later ;-).

Reply:
> According to my understanding there's no real gain in switching over the internal representation of available types from TypeDesc's to token

`READYTORUN_SECTION_AVAILABLE_TYPES` should be a secondary metadata index, nothing less or more. The most correct by construction way to create it is by walking the metadata directly.

I think it is pretty fragile to depend on the result of dependency analyzer to be always 100% identical to all types in the assembly here. I do not have a real example where they differ, but I can bet that one will show up eventually.